core/lib/zksync_core/src/eth_watch/mod.rs

```diff
@@ -178,27 +213,59 @@ impl EthWatch {
             .await?;
         stage_latency.observe();
 
+        let mut processed_events = 0;
         for processor in self.event_processors.iter_mut() {
-            processor
+            processed_events += processor
                 .process_events(storage, &*self.client, events.clone())
                 .await?;
         }
         self.last_processed_ethereum_block = to_block;
-        Ok(())
+        Ok(processed_events)
     }
 }
 
+/// This method creates an `EthWatch` instance that only looks for the `setChainId` event and runs
+/// it. Once the event is found and the tx is saved into the db, it will exit.
+pub async fn wait_for_set_chain_id(
+    config: ETHWatchConfig,
+    pool: ConnectionPool,
+    eth_gateway: Box<dyn EthInterface>,
+    diamond_proxy_addr: Address,
+    state_transition_manager_addr: Address,
+) -> anyhow::Result<()> {
+    let eth_client = EthHttpQueryClient::new(
+        eth_gateway,
+        diamond_proxy_addr,
+        Some(state_transition_manager_addr),
+        None,
+        config.confirmations_for_eth_event,
+    );
+
+    let mut eth_watch = EthWatch::new_set_chain_id_watch(
+        diamond_proxy_addr,
+        Box::new(eth_client),
+        config.poll_interval(),
+    )
+    .await;
+
+    let (_stop_sender, stop_receiver) = watch::channel(false);
```

Comment:
It feels strange that this method is non-responsive (i.e., doesn't accept `stop_receiver` from outside). Is this intentional? Is it assumed that it won't really need to wait for long realistically?

Reply:
Yes, this method is supposed to exit after a single `setChainId` event is found.